### PR TITLE
libfuzzer: experiment with improved strcmp TORCW handling

### PR DESCRIPTION
**Pull request by fmeum**
A (very small-scale) experiment with Jazzer has shown that libFuzzer's handling of string comparisons can be improved by adding the untruncated strings to the table of recent compares and using their actual lengths while mutating.

### CODE REVIEW

**Comment by fmeum** [comment]
The end of the logs for the failed CI job is:
```
2021-04-08T14:21:21.8145683Z Error response from daemon: Head https://registry-1.docker.io/v2/library/ubuntu/manifests/xenial: received unexpected HTTP status: 500 Internal Server Error
2021-04-08T14:21:21.8169169Z make: *** [docker/generated.mk:78: base-image] Error 1
2021-04-08T14:21:21.8683775Z ERROR:root:Retrying on run_command failed with Command '['make', 'RUNNING_ON_CI=yes', '-j', 'test-run-libfuzzer_vartorcw-jsoncpp_jsoncpp_fuzzer']' returned non-zero exit status 2.. Raise. Extras: {'traceback': 'Traceback (most recent call last):\n  File "/home/runner/work/fuzzbench/fuzzbench/common/retry.py", line 81, in _wrapper\n    result = func(*args, **kwargs)\n  File ".github/workflows/build_and_test_run_fuzzer_benchmarks.py", line 84, in run_command\n    subprocess.check_call(command)\n  File "/opt/hostedtoolcache/Python/3.8.8/x64/lib/python3.8/subprocess.py", line 364, in check_call\n    raise CalledProcessError(retcode, cmd)\nsubprocess.CalledProcessError: Command \'[\'make\', \'RUNNING_ON_CI=yes\', \'-j\', \'test-run-libfuzzer_vartorcw-jsoncpp_jsoncpp_fuzzer\']\' returned non-zero exit status 2.\n'}
2021-04-08T14:21:21.8692527Z Building fuzzer-benchmark pairs: [('libfuzzer_vartorcw', 'bloaty_fuzz_target'), ('libfuzzer_vartorcw', 'curl_curl_fuzzer_http'), ('libfuzzer_vartorcw', 'jsoncpp_jsoncpp_fuzzer'), ('libfuzzer_vartorcw', 'libpcap_fuzz_both'), ('libfuzzer_vartorcw', 'libxslt_xpath'), ('libfuzzer_vartorcw', 'mbedtls_fuzz_dtlsclient'), ('libfuzzer_vartorcw', 'openssl_x509'), ('libfuzzer_vartorcw', 'php_php-fuzz-parser'), ('libfuzzer_vartorcw', 'sqlite3_ossfuzz'), ('libfuzzer_vartorcw', 'systemd_fuzz-link-parser'), ('libfuzzer_vartorcw', 'zlib_zlib_uncompress_fuzzer')]
2021-04-08T14:21:21.8696270Z Running command: make RUNNING_ON_CI=yes -j test-run-libfuzzer_vartorcw-bloaty_fuzz_target
2021-04-08T14:21:21.8697908Z Running command: make RUNNING_ON_CI=yes -j test-run-libfuzzer_vartorcw-curl_curl_fuzzer_http
2021-04-08T14:21:21.8699593Z Running command: make RUNNING_ON_CI=yes -j test-run-libfuzzer_vartorcw-jsoncpp_jsoncpp_fuzzer
2021-04-08T14:21:21.8701298Z Running command: make RUNNING_ON_CI=yes -j test-run-libfuzzer_vartorcw-jsoncpp_jsoncpp_fuzzer
2021-04-08T14:21:21.8703075Z Running command: make RUNNING_ON_CI=yes -j test-run-libfuzzer_vartorcw-jsoncpp_jsoncpp_fuzzer
2021-04-08T14:21:21.8704002Z Traceback (most recent call last):
2021-04-08T14:21:21.8722419Z   File ".github/workflows/build_and_test_run_fuzzer_benchmarks.py", line 150, in <module>
2021-04-08T14:21:21.8723229Z     sys.exit(main())
2021-04-08T14:21:21.8724040Z   File ".github/workflows/build_and_test_run_fuzzer_benchmarks.py", line 145, in main
2021-04-08T14:21:21.8725066Z     result = do_build(build_type, fuzzer, always_build)
2021-04-08T14:21:21.8725932Z   File ".github/workflows/build_and_test_run_fuzzer_benchmarks.py", line 129, in do_build
2021-04-08T14:21:21.8727004Z     return make_builds(benchmarks, fuzzer)
2021-04-08T14:21:21.8727884Z   File ".github/workflows/build_and_test_run_fuzzer_benchmarks.py", line 98, in make_builds
2021-04-08T14:21:21.8728632Z     run_command(make_command)
2021-04-08T14:21:21.8729407Z   File "/home/runner/work/fuzzbench/fuzzbench/common/retry.py", line 81, in _wrapper
2021-04-08T14:21:21.8730137Z     result = func(*args, **kwargs)
2021-04-08T14:21:21.8730925Z   File ".github/workflows/build_and_test_run_fuzzer_benchmarks.py", line 84, in run_command
2021-04-08T14:21:21.8731748Z     subprocess.check_call(command)
2021-04-08T14:21:21.8732656Z   File "/opt/hostedtoolcache/Python/3.8.8/x64/lib/python3.8/subprocess.py", line 364, in check_call
2021-04-08T14:21:21.8733631Z     raise CalledProcessError(retcode, cmd)
2021-04-08T14:21:21.8735654Z subprocess.CalledProcessError: Command '['make', 'RUNNING_ON_CI=yes', '-j', 'test-run-libfuzzer_vartorcw-jsoncpp_jsoncpp_fuzzer']' returned non-zero exit status 2.
2021-04-08T14:21:21.9618055Z ##[error]Process completed with exit code 1.
```

It seems that this failure is caused by a Docker registry problem, but I'm not sure that is what caused the fatal error.

File: fuzzers/libfuzzer_vartorcw/builder.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+ARG parent_image
+FROM $parent_image
+
+RUN git clone https://github.com/CodeIntelligenceTesting/llvm-project.git /llvm-project && \
```
Comment (inferno-chromium) [inline fuzzers/libfuzzer_vartorcw/builder.Dockerfile:18]: Does this have any other custom changes, can we just use the main llvm-project ?

**Review by inferno-chromium — COMMENTED** [review]
If you push again, that retriggers CI and hopefully it will overcome flake.

File: fuzzers/libfuzzer_vartorcw/builder.Dockerfile (hunk 1 above)
Comment (fmeum) [reply fuzzers/libfuzzer_vartorcw/builder.Dockerfile:18]: This is checkout is referencing [this commit](https://github.com/CodeIntelligenceTesting/llvm-project/commit/bbe1270aebdbd966d027d93c4a876499ac1573a9) in the CodeIntelligenceTesting/llvm-project repo. That commit contains my changes and its parent is the commit currently used by the `libfuzzer` core fuzzer.

**Comment by fmeum** [comment]
> If you push again, that retriggers CI and hopefully it will overcome flake.

It did, all checks pass now.

**Review by inferno-chromium — APPROVED** [review]

**Pull request merged**